### DISCUSSION TITLE
$metadata missing on dynamically registered models

### DISCUSSION THREAD

**Discussion by Remo** [OData/AspNetCoreOData#1318] [Q&A] [answered]
I'm using a rather dynamic way to register my models. It's a a bit like shown here: https://github.com/xuzhg/WebApiSample/tree/main/v8.x/DynamicRouteSample

It's all working well till I want to get my meta information.

The linked project has the same problem.
I can query the data with something like this: http://localhost:5182/v1/people

This will then return the following JSON:

```json
{
    "@odata.context": "http://localhost:5182/v1/$metadata#People",
    "value": [
        {
            "Id": 1,
            "Name": "Chilly",
            "Emails": [
                "Earth@abc.com",
                "Xia@st.com"
            ]
        },
        {
            "Id": 2,
            "Name": "Mild",
            "Emails": [
                "Pat@abc.com",
                "101@YE.com"
            ]
        },
        {
            "Id": 3,
            "Name": "Balmy",
            "Emails": [
                "GG@abc.com",
                "1241@WU.com"
            ]
        },
        {
            "Id": 4,
            "Name": "Hot",
            "Emails": [
                "Ket@abc.com",
                "14@YU.com"
            ]
        }
    ]
}
```

As you can see, the odata context is http://localhost:5182/v1/$metadata#People, but when I try to open this, I only get a 404 response. I'm a bit lost, not sure where to start looking. 

* Copilot suggested I add `.AddOData(o => o.EnableQueryFeatures());`, but no luck
* Google also didn't help
* Nothing relevant in this github project

**Comment by habbes** [comment]
Hello @Remo, I'm not sure I understand the problem. What is that you try to open that returns a 404 response? You've said that the endpoint `http://localhost:5182/v1/people` returns the JSON response you've shown above, where does the 404 come from?

**Reply by Remo** [reply]
@habbes I'm trying to see the metadata: http://localhost:5182/v1/$metadata#People The link listed in the JSON

**Reply by habbes** [reply, marked as answer]
@Remo thanks for the clarification. I had misunderstood the message. The `/$metadata` endpoint returns the schema of the service as defined in your `IEdmModel`. This endpoint is usually handled by a built-in controller that's registered when you call `AddOData()`.

However, in the project that you have linked it seems it uses custom dynamic rooting, and not the built-in one. I have not gone through the entire codebase to see what could be missing, but an easy workaround would be to add a custom route in your controller to handle the `$metadata` endpoint and return the model:

```c#
public class AnyNameForController : Controller
{
    [HttpGet]
    [ODataRoute("$metadata")]
    public IEdmModel GetMetadata()
    {
        IODataFeature feature = Request.ODataFeature();
        string prefix = feature.RoutingConventionsStore["odata_prefix"] as string;
        
        return prefix == "v1" ? EdmModelBuilder.GetEdmModel1() : EdmModelBuilder.GetEdmModel2();
    }

    // ...
}
```

**Reply by Remo** [reply]
Thanks @habbes! Very much appreciated, it works.
Thanks @xuzhg Thanks for changing your sample so quickly. For reference https://github.com/xuzhg/WebApiSample/commit/3836b45d5db53bb94eccdf0e7121da5cef1a20c4
